src/ecs/database/create-database.ts

```diff
@@ -276,8 +276,12 @@ export function createDatabase<
         resources,
         transactions,
         observe,
-        toData: () => store.toData(),
+        toData: () => {
+            store.compact();
+            return store.toData();
+        },
         fromData: (data: unknown) => {
+            store.compact();
```

Comment:
Let's not compact on fromData, especially right before reading in data, that would accomplish nothing.